File: dong99u/leetcode/3640.java
package leetcode.leetcode3640;

import java.util.*;

/**
 * 4 <= n <= 10^5 -> 최대 O(nlogn)
 * 이 문제를 brute force로 풀면 O(n^5)?? -> O(n^2)만큼 [l, r] 구간 잡고, O(n^2)만큼 [p, q] 구간 잡아서 O(n)만큼 단조 증가/감소 확인
 * dp
 */
public class Solution {
	static final long MIN_VALUE = -Long.MAX_VALUE;

	public static void main(String[] args) {
		System.out.println(maxSumTrionic(new int[] {0, -2, -1, -3, 0, 2, -1}));
	}
	public static long maxSumTrionic(int[] nums) {
		int n = nums.length;
		long[][] dp = new long[n][4];

		// dp 테이블을 -inf로 초기화
		for (int i = 0; i < n; i++) {
			Arrays.fill(dp[i], MIN_VALUE);
		}

		/**
		 * 조건 충족 시 (k=1,3이면 arr[i-1] < arr[i], k=2면 arr[i-1] > arr[i]):
		 *
		 *   dp[i][k] = max(
		 *       dp[i-1][k]   + arr[i],   // ① 확장 (같은 단계 안에서 한 칸 연장)
		 *       dp[i-1][k-1] + arr[i]    // ② 전이 (이전 단계에서 막 넘어옴)
		 *   )
		 */

		for (int i = 1; i < n; i++) {
			boolean isIncrement = nums[i - 1] < nums[i];
			boolean isDecrement = nums[i - 1] > nums[i];

			// k = 1: 증가 단계
			if (isIncrement) {
				// 확장 (같은 단계 안에서 한 칸 연장)
				long curr = (dp[i - 1][1] != MIN_VALUE) ? dp[i - 1][1] + nums[i] : MIN_VALUE;
				// 새로 시작 (길이 2짜리 증가 부분배열로 출발)
				long next = nums[i - 1] + nums[i];
				dp[i][1] = Math.max(curr, next);
			}

			if (isDecrement) {
				// 확장 (같은 단계 안에서 한 칸 연장)
				long curr = (dp[i - 1][2] != MIN_VALUE) ? dp[i - 1][2] + nums[i] : MIN_VALUE;
				// 전이 (이전 단계에서 막 넘어옴)
				long next = (dp[i-1][1] != MIN_VALUE) ? dp[i-1][1] + nums[i] : MIN_VALUE;
				dp[i][2] = Math.max(curr, next);
			}

			if (isIncrement) {
				long curr = (dp[i - 1][3] != MIN_VALUE) ? dp[i - 1][3] + nums[i] : MIN_VALUE;
				long next = (dp[i - 1][2] != MIN_VALUE) ? dp[i - 1][2] + nums[i] : MIN_VALUE;
				dp[i][3] = Math.max(curr, next);
			}
		}

		long answer = MIN_VALUE;
		for (int i = 0; i < n; i++) {
			answer = Math.max(answer, dp[i][3]);
		}
		return answer;
    }
}

File: jiseok/LeetCode/3640.py
class Solution:
    def maxSumTrionic(self, nums: List[int]) -> int:
        # dp 문제 같은데
        N = len(nums)
        in_or_decrease = [0 for _ in range(N)] # 현재 인덱스 기준에서 다음이 감소면 -1, 증가면 1, 그대로면 0
        for i in range(N - 1):
            if nums[i] > nums[i + 1]:
                in_or_decrease[i] = -1
            elif nums[i] < nums[i + 1]:
                in_or_decrease[i] = 1

        # 0이 나오면 무조건 처음부터 
        # 시작 인덱스 기억해야할 듯
        # 현재 상태도 저장-> 증가 중(첫 번째)인지 감소 중인지 증가 중(두 번째)인지
        cur_state = 0
        start_idx = -1
        dp = [float('-inf') for _ in range(N)]
        for i in range(N - 1):
            if cur_state == 0:
                # 지금까지의 상태가 유지일 때 다음이 증가한다면 상태는 1로 변경하고 시작 인덱스 설정
                if in_or_decrease[i] == 1:
                    cur_state = 1
                    start_idx = i

            # 지금까지의 상태가 증가 중(첫 번째)일 때
            elif cur_state == 1:
                # 다음이 감소라면
                if in_or_decrease[i] == -1:
                    cur_state = 2
                elif in_or_decrease[i] == 0:
                    cur_state = 0

            # 지금까지의 상태가 감소 중일 때
            elif cur_state == 2:
                if in_or_decrease[i] == 1:
                    cur_state = 3
                    # 오답의 이유가 second_start_idx 같은데 어떻게 처리해야 할지 모르겠음
                    second_start_idx = i
                elif in_or_decrease[i] == 0:
                    cur_state = 0

            # 지금까지의 상태가 증가 중(두 번째)일 때
            else:
                if in_or_decrease[i] == -1:
                    cur_state = 2
                    start_idx = second_start_idx
                elif in_or_decrease[i] == 0:
                    cur_state = 0
            
            if cur_state == 3:
                dp[i] = sum(nums[start_idx : i + 2])

        return max(dp)
